python/ray/autoscaler/v2/scheduler.py

```diff
@@ -0,0 +1,302 @@
+    TO_LAUNCH = "TO_LAUNCH"
+    # The node is pending to be launched (a pending request, or a pending instance)
+    PENDING = "PENDING"
+    # The node is running.
+    RUNNING = "RUNNING"
+
+
+@dataclass
+class SchedulingNode:
+    """
+    A abstraction of a node that can be scheduled on by the resource scheduler.
+
+    A scheduling node is expected to be used as:
+
+        node  = SchedulingNode.from_node_config(node_config)
+        infeasible = node.try_schedule(requests)
+        score = node.compute_score()
+
+        .... do something with the score ....
+
+    NOTE:
+        One could also extend the scheduling behavior by overriding:
+            1. try_schedule()
+            2. compute_score()
+
+    """
+
+    # Class level node id counter.
+    _next_node_id = 0
+
+    # A unique id simply for differentiating nodes. Not the same as a normal ray node
+    # id.
+    node_id: int
+    # Node type name.
+    node_type: NodeType
+    # Requests committed to be placed on this node.
+    sched_requests: List[ResourceRequest] = field(default_factory=list)
+    # The node's current resource capacity.
+    total_resources: Dict[str, float] = field(default_factory=dict)
+    # The node's available resources.
+    available_resources: Dict[str, float] = field(default_factory=dict)
+    # Node's labels, including static or dynamic labels.
+    labels: Dict[str, str] = field(default_factory=dict)
+    # Status
+    status: SchedulingNodeStatus = SchedulingNodeStatus.TO_LAUNCH
+    # Observability descriptive message for why the node was launched in the
+    # first place.
+    # TODO
+    launch_reason: Optional[str] = None
+
+    @classmethod
+    def from_node_config(
+        cls,
+        node_config: NodeTypeConfig,
+        status: SchedulingNodeStatus = SchedulingNodeStatus.TO_LAUNCH,
+    ) -> "SchedulingNode":
+        """
+        Create a scheduling node from a node config.
+        """
+        return cls(
+            node_id=cls.next_node_id(),
+            node_type=node_config.name,
+            total_resources=dict(node_config.resources),
+            available_resources=dict(node_config.resources),
+            labels=dict(node_config.labels),
+            status=status,
+        )
+
+    def try_schedule(self, requests: List[ResourceRequest]) -> List[ResourceRequest]:
+        """
+        Try to schedule the resource requests on this node.
+
+        This modifies the node's available resources if the requests are schedulable.
+        When iterating through the requests, the requests are sorted by the
+        `_sort_resource_request` function. The requests are scheduled one by one in
+        the sorted order, and no backtracking is done.
+
+        Args:
+            requests: The resource requests to be scheduled.
+
+        Returns:
+            A list of infeasible requests that cannot be scheduled on this node.
+        """
+        pass
+
+    def compute_score(self) -> UtilizationScore:
```

Comment:
where is the `current resource request`?

Reply:
It will be reflected on the `available_resources` indirectly - we don't know what requests are being scheduled on a running node.